{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "5bc1187c",
   "metadata": {},
   "outputs": [],
   "source": [
    "import os\n",
    "import json\n",
    "\n",
    "\n",
    "import numpy as np\n",
    "import pandas as pd"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "24551db5",
   "metadata": {},
   "outputs": [],
   "source": [
    "data_path = '/home/hmbaghda/orcd/pool/metastatic_potential/'\n",
    "random_state = 42 + 3\n",
    "\n",
    "n_cores = 30\n",
    "os.environ[\"OMP_NUM_THREADS\"] = str(n_cores)\n",
    "os.environ[\"MKL_NUM_THREADS\"] = str(n_cores)\n",
    "os.environ[\"OPENBLAS_NUM_THREADS\"] = str(n_cores)\n",
    "os.environ[\"VECLIB_MAXIMUM_THREADS\"] = str(n_cores)\n",
    "os.environ[\"NUMEXPR_NUM_THREADS\"] = str(n_cores)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "cceeab26",
   "metadata": {},
   "outputs": [],
   "source": [
    "X = pd.read_csv(os.path.join(data_path, 'processed',  'expr_joint.csv'), index_col = 0)\n",
    "expr_joint = X.copy()\n",
    "\n",
    "mp_joint=pd.read_csv(os.path.join(data_path, 'processed', 'metastatic_potential_joint.csv'), index_col = 0)['mean']\n",
    "y = mp_joint.values.ravel()\n",
    "\n",
    "expr_protein = pd.read_csv(os.path.join(data_path, 'processed',  'expr_protein.csv'), index_col = 0)\n",
    "expr_rna = pd.read_csv(os.path.join(data_path, 'processed',  'expr.csv'), index_col = 0)\n",
    "\n",
    "protein_cols = expr_protein.columns\n",
    "rna_cols = expr_rna.columns\n",
    "\n",
    "# with open(\"protein_cols.txt\", \"w\") as f: f.writelines(f\"{item}\\n\" for item in protein_cols)\n",
    "# with open(\"rna_cols.txt\", \"w\") as f: f.writelines(f\"{item}\\n\" for item in rna_cols)\n",
    "\n",
    "\n",
    "X_protein = X[protein_cols].values\n",
    "X_rna = X[rna_cols].values"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "ca686b4c",
   "metadata": {},
   "source": [
    "ID mapping between protein and RNA:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "abc16067",
   "metadata": {},
   "outputs": [],
   "source": [
    "with open(os.path.join(data_path, 'processed', 'uniprot_mapper.json'), \"r\") as json_file:\n",
    "    uid_mapper = json.load(json_file)\n",
    "    \n",
    "# manually mapped some that failed to map using uniprot ID\n",
    "manual_map = {'Q9TNN7': 'HLA-C',\n",
    "'P16189': 'HLA-A',\n",
    "'P30456': 'HLA-A',\n",
    "'P30443': 'HLA-A',\n",
    "'P05534': 'HLA-A',\n",
    "'P18462': 'HLA-A',\n",
    "'P01892': 'HLA-A',\n",
    "'P13746': 'HLA-A',\n",
    "'P01891': 'HLA-A',\n",
    "'P30483': 'HLA-B',\n",
    "'P30484': 'HLA-B',\n",
    "'P03989': 'HLA-B',\n",
    "'P30460': 'HLA-B',\n",
    "'P30461': 'HLA-B',\n",
    "'Q95365': 'HLA-B',\n",
    "'P16188': 'HLA-A',\n",
    "'Q95604': 'HLA-C',\n",
    "'Q07000': 'HLA-C',\n",
    "'P30499': 'HLA-C',\n",
    "'P30501': 'HLA-C',\n",
    "'P30504': 'HLA-C',\n",
    "'Q95IE3': 'HLA-DRB1',\n",
    "'P04229': 'HLA-DRB1',\n",
    "'P20039': 'HLA-DRB1',\n",
    "'P13760': 'HLA-DRB1',\n",
    "'Q5Y7A7': 'HLA-DRB1',\n",
    "'Q9GIY3': 'HLA-DRB1',\n",
    "'Q9TQE0': 'HLA-DRB1',\n",
    "'Q30134': 'HLA-DRB1'}\n",
    "    \n",
    "protein_names = []\n",
    "for protein_id in protein_cols:\n",
    "    uniprot_id = protein_id.split('|')[1].split('-')[0]\n",
    "    if pd.isna(uid_mapper[uniprot_id]):\n",
    "        gene_name = protein_id.split('|')[-1].split('_HUMAN')[0]\n",
    "        if gene_name[0].isdigit():\n",
    "            gene_name = manual_map[uniprot_id]\n",
    "    else:\n",
    "        gene_name = uid_mapper[uniprot_id]\n",
    "    protein_names.append(gene_name)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "be712f83",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Of the 19138 RNA features and 10969 protein features, there are 10291 features in common\n"
     ]
    }
   ],
   "source": [
    "rna_names = [rna_id.split(' (')[0] for rna_id in rna_cols]\n",
    "# protein_names = [protein_id.split('|')[-1].split('_HUMAN')[0] for protein_id in protein_cols]\n",
    "intersect_names = set(rna_names).intersection(protein_names)\n",
    "\n",
    "n_features = [len(rna_names), len(protein_names), len(intersect_names)]\n",
    "print('Of the {} RNA features and {} protein features, there are {} features in common'.format(*n_features))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "983644a1",
   "metadata": {},
   "outputs": [],
   "source": [
    "expr_protein.columns = protein_names\n",
    "expr_rna.columns = rna_names"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "id": "d06237bc",
   "metadata": {},
   "outputs": [],
   "source": [
    "import json\n",
    "\n",
    "prot_id_map = dict(zip(protein_cols.tolist(), protein_names))\n",
    "rna_id_map = dict(zip(rna_names, rna_cols.tolist()))\n",
    "modality_id_map = {k: rna_id_map[v] for k,v in prot_id_map.items() if v in intersect_names}\n",
    "\n",
    "with open(os.path.join(data_path, 'processed', 'modality_mapper.json'), \"w\") as json_file:\n",
    "    json.dump(modality_id_map, json_file)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "f8366dd0",
   "metadata": {},
   "source": [
    "Now, any column that is shared between expr_protein and expr_rna is the same gene product. \n",
    "\n",
    "Modality map creates a direct mapping between the proteomic feature name and transcriptomic feature name. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "35ded73c",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python [conda env:metastatic_potential]",
   "language": "python",
   "name": "conda-env-metastatic_potential-py"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.14.0"
  },
  "toc": {
   "base_numbering": 1,
   "nav_menu": {},
   "number_sections": true,
   "sideBar": true,
   "skip_h1_title": false,
   "title_cell": "Table of Contents",
   "title_sidebar": "Contents",
   "toc_cell": false,
   "toc_position": {},
   "toc_section_display": true,
   "toc_window_display": false
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
